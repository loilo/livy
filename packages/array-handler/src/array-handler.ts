import type {
  LogRecord,
  ResettableInterface,
  SyncHandlerInterface,
} from '@livy/contracts'
import { AbstractLevelBubbleHandler } from '@livy/util/handlers/abstract-level-bubble-handler'
import { MirrorSyncHandlerMixin } from '@livy/util/handlers/mirror-sync-handler-mixin'
import { ProcessableHandlerMixin } from '@livy/util/handlers/processable-handler-mixin'

/**
 * Stores log records in an array; great for debugging
 */
export class ArrayHandler
  extends MirrorSyncHandlerMixin(
    ProcessableHandlerMixin(AbstractLevelBubbleHandler),
  )
  implements SyncHandlerInterface, ResettableInterface
{
  private _records: LogRecord[] = []

  /**
   * @inheritdoc
   */
  public handleSync(record: LogRecord) {
    this._records.push(record)
    return !this.bubble
  }

  /**
   * @inheritdoc
   */
  public handleBatchSync(records: LogRecord[]) {
    this._records.push(...records)
  }

  /**
   * Get the stored records
   */
  public get records() {
    return this._records
  }

  /**
   * @inheritdoc
   */
  public reset() {
    this._records.splice(0)
  }
}
