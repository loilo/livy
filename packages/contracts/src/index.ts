export * from './closable-handler-interface.js'
export * from './formattable-handler-interface.js'
export * from './formatter-interface.js'
export * from './handler-interface.js'
export * from './log-level.js'
export * from './log-record.js'
export * from './logger-interface.js'
export * from './processable-handler-interface.js'
export * from './processor-interface.js'
export * from './resettable-interface.js'
