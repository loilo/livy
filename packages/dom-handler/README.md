# `@livy/dom-handler`

This [Livy](../../README.md#readme) handler writes log records to the browser's DOM.

---

[**Synchronous logger support:**](../../README.md#synchronous-and-asynchronous-logging) yes

**Runtime:** [Browsers](../../README.md#usage-in-browsers)

---

## Basic Example

```js
import { DomHandler } from '@livy/dom-handler'

const handler = new DomHandler('#logs')
```

## Installation

Install it via npm:

```bash
npm install @livy/dom-handler
```

## Options

The first argument to this handler's constructor is a selector string or a [DOM element](https://developer.mozilla.org/en-US/docs/Web/API/Element) to write the log records to.

> **Note:** If you pass a selector string, it will immediately try to find it in the DOM tree. If no results are yielded, the selector will be looked up again once the DOM [finished loading](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event). If no according element is found by then, the handler will throw an error.
>
> If you're running your logger in [async mode](../../README.md#synchronous-and-asynchronous-logging), logged records will simply be deferred until that point. However, if you're logging in mixed or sync mode, the logger will throw an error if no target element has been found yet.

An object of options can be passed to the constructor as the second argument.

The following options are available:

### `autoScroll`

**Type:** `'edge' | 'force' | 'none'`

**Default:** `'edge'`

**Description:** Scroll behavior when new log records are added:

- `edge`: scroll new entry into view if container has previously been scrolled to the edge
- `force`: always scroll new entries into view
- `none`: no automatic scrolling

### `bubble`

**Type:** `boolean`

**Default:** `true`

**Description:** Controls whether records handled by this handler should bubble up to other handlers.

> See also: [Bubbling](../../README.md#bubbling)

### `formatter`

**Type:** [`FormatterInterface`](../contracts/README.md#formatterinterface)

**Default:** [`new HtmlPrettyFormatter()`](../html-pretty-formatter/README.md#readme)

**Description:** The formatter to use.

### `level`

**Type:** [`LogLevel`](../contracts/README.md#loglevel)

**Default:** `'debug'`

**Description:** Controls which log records should be handled based on their [log level](../../README.md#log-levels).

### `reversed`

**Type:** `boolean`

**Default:** `false`

**Description:** Whether DOM elements should be added at the beginning of the container instead of the end

## Public API

### `bubble`

Controls whether records handled by this handler should bubble up to other handlers. Initially set through the [`bubble`](#bubble) option.

> See also: [Bubbling](../../README.md#bubbling)

### `close()`

This handler implements the [`ClosableHandlerInterface`](../contracts/README.md#closablehandlerinterface) and does internal cleanup work.

> You usually don't want to call this method manually. It is done automatically when a Node.js process exits / a browser page is closed.

### `defaultFormatter` (read-only)

The formatter used by this handler if no [`formatter`](#formatter) option is set.

### `formatter`

This handler supports formatters by implementing the [`FormattableHandlerInterface`](../contracts/README.md#formattablehandlerinterface).

### `level`

The minimum [log level](../../README.md#log-levels) of a log record to be considered by this handler. Initially set through the [`level`](#level) option.

### `processors`

This handler supports [processors](../../README.md#processors) by implementing the [`ProcessableHandlerInterface`](../contracts/README.md#processablehandlerinterface).

### `reset()`

This handler implements the [`ResettableInterface`](../contracts/README.md#resettableinterface). Resetting it resets all attached processors.

> You usually don't want to call this method manually on an individual handler. Consider calling it [on the logger](../logger/README.md#reset) instead.
