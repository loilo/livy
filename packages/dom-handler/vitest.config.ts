import { defineConfig } from 'vitest/config'
import { vitestConfig } from '@livy/test-utils/vitest-config.js'
import { defu } from 'defu'

export default defineConfig(
  defu(
    {
      test: {
        environment: 'jsdom'
      }
    },
    vitestConfig
  )
)
