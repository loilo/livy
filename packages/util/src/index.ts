export type * from './formatters/abstract-batch-formatter.js'
export type * from './formatters/included-record-properties.js'
export type * from './formatters/line-formatter.js'
export type * from './handlers/abstract-batch-handler.js'
export type * from './handlers/abstract-formatting-processing-handler.js'
export type * from './handlers/abstract-level-bubble-handler.js'
export type * from './handlers/formattable-handler-mixin.js'
export type * from './handlers/is-closable-handler-interface.js'
export type * from './handlers/is-formattable-handler-interface.js'
export type * from './handlers/is-processable-handler-interface.js'
export type * from './handlers/is-sync-handler-interface.js'
export type * from './handlers/mirror-sync-handler-mixin.js'
export type * from './handlers/processable-handler-mixin.js'
export type * from './handlers/respect-level-mixin.js'
export type * from './console-adapter-interface.js'
export type * from './console-adapter.js'
export type * from './environment.js'
export type * from './gated-set.js'
export type * from './helpers.js'
export type * from './html.js'
export type * from './is-resettable-interface.js'
export type * from './mixin.js'
export type * from './timer.js'
export type * from './types.js'
export type * from './validatable-set.js'
