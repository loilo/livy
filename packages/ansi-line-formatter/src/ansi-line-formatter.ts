import type { LogLevel } from '@livy/contracts'
import { LineFormatter, LineFormatterOptions } from '@livy/line-formatter'
import { isEmpty } from '@livy/util/helpers'
import chalk, { supportsColor } from 'chalk'
import { DateTime } from 'luxon'

import { emphasize } from 'emphasize'
import chalkTemplate from 'chalk-template'

export interface AnsiLineFormatterOptions extends LineFormatterOptions {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   * If not provided (or set to `undefined`), color support is going
   * to be automatically detected
   */
  decorated: boolean | undefined
}

/**
 * Formats log records as single lines with terminal highlighting
 */
export class AnsiLineFormatter extends LineFormatter {
  /**
   * Whether records should be highlighted with ANSI escape sequences
   */
  public decorated: boolean | undefined

  public constructor({
    decorated,
    ...options
  }: Partial<AnsiLineFormatterOptions> = {}) {
    super(options)
    this.decorated = decorated
  }

  /**
   * @inheritdoc
   */
  protected formatDatetime(datetime: DateTime) {
    const timestamp = super.formatDatetime(datetime)

    if (this.shouldDecorate()) {
      return chalk.dim(timestamp)
    } else {
      return timestamp
    }
  }

  /**
   * @inheritdoc
   */
  protected formatLevel(level: LogLevel) {
    const formatted = super.formatLevel(level)

    if (!this.shouldDecorate()) {
      return formatted
    } else {
      let color: string
      switch (level) {
        case 'emergency':
          color = 'red'
          break

        case 'alert':
          color = 'red'
          break

        case 'critical':
          color = 'red'
          break

        case 'error':
          color = 'red'
          break

        case 'warning':
          color = 'yellow'
          break

        case 'notice':
          color = 'blue'
          break

        case 'info':
          color = 'blue'
          break

        case 'debug':
          color = 'blue.dim'
          break

        /* c8 ignore next 2: This should never happen, but is included for type safety */
        default:
          return formatted
      }

      return chalkTemplate`{${color} ${formatted}}`
    }
  }

  /**
   * @inheritdoc
   */
  protected formatData(object: any, ignoreEmpty: boolean) {
    if (isEmpty(object) && ignoreEmpty) {
      return ''
    }

    const stringified = super.formatData(object)

    if (this.shouldDecorate()) {
      return emphasize.highlight('json', stringified).value
    } else {
      return stringified
    }
  }

  /**
   * Check whether the formatter should use ANSI codes to decorate log records
   */
  public shouldDecorate() {
    return Boolean(
      this.decorated !== undefined
        ? this.decorated
        : supportsColor
        ? supportsColor.hasBasic
        : false,
    )
  }
}
