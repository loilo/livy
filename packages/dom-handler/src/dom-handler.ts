import type {
  ClosableHandlerInterface,
  FormatterInterface,
  LogRecord,
  ResettableInterface,
} from '@livy/contracts'
import { HtmlPrettyFormatter } from '@livy/html-pretty-formatter'
import { AbstractSyncFormattingProcessingHandler } from '@livy/util/handlers/abstract-formatting-processing-handler'
import { AbstractLevelBubbleHandlerOptions } from '@livy/util/handlers/abstract-level-bubble-handler'
import { isPromiseLike } from '@livy/util/helpers'
import { Promisable } from '@livy/util/types'

export type AutoScroll = 'edge' | 'force' | 'none'

export interface DomHandlerOptions extends AbstractLevelBubbleHandlerOptions {
  /**
   * Scroll behavior when new entries are inserted:
   * - edge: scroll new entry into view if container has previously been scrolled to the edge
   * - force: always scroll new entries into view
   * - none: no automatic scrolling
   */
  autoScroll: AutoScroll

  /**
   * The formatter to use
   */
  formatter: FormatterInterface

  /**
   * Whether DOM elements should be added at the beginning of the container instead of the end
   */
  reversed: boolean
}

/**
 * Writes log records to the browser DOM
 */
export class DomHandler
  extends AbstractSyncFormattingProcessingHandler
  implements ResettableInterface, ClosableHandlerInterface
{
  public reversed: boolean
  public autoScroll: AutoScroll
  private readyHandler?: () => void

  /**
   * The DOM element to attach log entries to
   */
  protected container: Promisable<Element>

  public constructor(
    container: string | Element,
    {
      reversed = false,
      autoScroll = 'edge',
      formatter,
      ...options
    }: Partial<DomHandlerOptions> = {},
  ) {
    super(options)

    this.autoScroll = autoScroll
    this.reversed = reversed
    this.explicitFormatter = formatter

    if (typeof container === 'string') {
      const possibleContainer = document.querySelector(container)
      if (possibleContainer) {
        this.container = possibleContainer
      } else {
        this.container = this.watchDomReadyState().then(() => {
          const queriedContainerElement = document.querySelector(container)

          if (queriedContainerElement === null) {
            throw new Error(
              `Could not find DomHandler target element at selector "${container}"`,
            )
          }

          this.container = queriedContainerElement

          return queriedContainerElement
        })
      }
    } else {
      this.container = container
    }
  }

  /**
   * @inheritdoc
   */
  public getDefaultFormatter(): FormatterInterface {
    return new HtmlPrettyFormatter()
  }

  /**
   * Get a promise that resolves when the DOM is ready
   */
  private watchDomReadyState(): Promise<void> {
    if (document.readyState === 'loading') {
      return new Promise(resolve => {
        this.readyHandler = () => {
          if (document.readyState !== 'loading') {
            document.removeEventListener('DOMContentLoaded', this.readyHandler!)
            this.readyHandler = undefined
            resolve()
          }
        }

        document.addEventListener('DOMContentLoaded', this.readyHandler)
      })
    } else {
      /* c8 ignore next: This ist just a fallback that should usually not occur in practice */
      return Promise.resolve()
    }
  }

  /**
   * Perform synchronous or asynchrounus writing
   *
   * @param formatted The formatted string to write
   * @param container The container element the formatted entries go to
   */
  private doWrite(formatted: string, container: Element) {
    let isAtEdge = false

    /* c8 ignore start: Scrolling cannot be tested with the JSDOM mock */
    if (this.autoScroll === 'force') {
      isAtEdge = true
    } else if (this.autoScroll === 'edge') {
      isAtEdge = this.reversed
        ? container.scrollTop === 0
        : container.scrollTop ===
          container.scrollHeight - container.clientHeight
    }
    /* c8 ignore stop */

    container.insertAdjacentHTML(
      this.reversed ? 'afterbegin' : 'beforeend',
      formatted,
    )

    /* c8 ignore start: Scrolling cannot be tested with the JSDOM mock */
    if (isAtEdge) {
      container.scrollTop = this.reversed
        ? 0
        : container.scrollHeight - container.clientHeight
    }
    /* c8 ignore stop */
  }

  /**
   * @inheritdoc
   */
  protected async write(_record: LogRecord, formatted: string) {
    const container = await this.container
    return this.doWrite(formatted, container)
  }

  /**
   * @inheritdoc
   */
  protected writeSync(_record: LogRecord, formatted: string) {
    const container = this.container

    if (isPromiseLike(container)) {
      throw new Error(`Could not find DomHandler target element`)
    }

    return this.doWrite(formatted, container)
  }

  /**
   * @inheritdoc
   */
  public reset() {
    if (this.container instanceof Element) {
      this.container.innerHTML = ''
    }
  }

  /**
   * @inheritdoc
   */
  public close() {
    if (this.readyHandler) {
      document.removeEventListener('DOMContentLoaded', this.readyHandler)
      this.readyHandler = undefined
    }
  }
}
