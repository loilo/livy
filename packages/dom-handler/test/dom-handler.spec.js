import { describe, expect, afterEach, it } from 'vitest'
import { DomHandler } from '../src/dom-handler'

const { record } = livyTestGlobals

describe('@livy/dom-handler', () => {
  afterEach(() => {
    document.body.innerHTML = ''
  })

  it('should write logs to the referenced element', () => {
    const container = document.createElement('div')
    const handler = new DomHandler(container)
    handler.handleSync(record('debug', 'Test DomHandler'))
    handler.handleSync(record('info', 'Test DomHandler'))

    expect(container.innerHTML).toMatchSnapshot()
  })

  it('should write logs to a selector (with DOM ready)', () => {
    const container = document.createElement('div')
    container.id = 'logs'
    document.body.append(container)

    const handler = new DomHandler('#logs')
    handler.handleSync(record('debug', 'Test DomHandler'))

    expect(container.innerHTML).toMatchSnapshot()
  })

  describe('interaction with "loading" document.readyState', () => {
    let handler

    afterEach(() => {
      handler.close()
      delete document.readyState
    })

    it('should asynchronously write logs to a selector (with DOM not initially ready)', async () => {
      const container = document.createElement('div')
      container.id = 'logs'

      // Override document.readyState to be "loading"
      Object.defineProperty(document, 'readyState', {
        value: 'loading',
        writable: true,
        configurable: true
      })

      handler = new DomHandler('#logs')
      const logged = handler.handle(record('debug', 'Test DomHandler'))

      // Attach element, update document.readyState, dispatch event
      document.body.append(container)
      document.readyState = 'complete'
      document.dispatchEvent(new Event('DOMContentLoaded'))

      await logged
      expect(container.innerHTML).toMatchSnapshot()
    })

    it('should fail to asynchronously write logs to a non-existing selector (with DOM not initially ready)', async () => {
      const container = document.createElement('div')
      container.id = 'logs'

      // Override document.readyState to be "loading"
      Object.defineProperty(document, 'readyState', {
        value: 'loading',
        writable: true,
        configurable: true
      })

      handler = new DomHandler('#logs')
      const logged = handler.handle(record('debug', 'Test DomHandler'))

      // Attach element, update document.readyState, dispatch event
      document.readyState = 'complete'
      document.dispatchEvent(new Event('DOMContentLoaded'))

      let notFoundError
      try {
        await logged
      } catch (error) {
        notFoundError = error
      }

      expect(notFoundError).toEqual(
        new Error(
          `Could not find DomHandler target element at selector "#logs"`
        )
      )
    })

    it('should fail to synchronously write logs to a non-existing selector (with DOM not ready)', () => {
      const container = document.createElement('div')
      container.id = 'logs'

      // Override document.readyState to be "loading"
      Object.defineProperty(document, 'readyState', {
        value: 'loading',
        writable: true,
        configurable: true
      })

      handler = new DomHandler('#logs')

      document.body.appendChild(container)

      expect(() => {
        handler.handleSync(record('debug', 'Test DomHandler'))
      }).toThrowError(new Error(`Could not find DomHandler target element`))
    })
  })

  it('should clear the DOM element on reset', () => {
    const container = document.createElement('div')

    const handler = new DomHandler(container)
    handler.handleSync(record('info', 'Test ArrayHandler'))
    handler.reset()

    expect(container.innerHTML).toBe('')
  })

  it('should respect the "reversed" option', () => {
    const container = document.createElement('div')
    const handler = new DomHandler(container, { reversed: true })
    handler.handleSync(record('debug', 'Test DomHandler'))
    handler.handleSync(record('info', 'Test DomHandler'))

    expect(container.innerHTML).toMatchSnapshot()
  })

  it('should respect the "formatter" option', () => {
    const container = document.createElement('div')
    const handler = new DomHandler(container, {
      formatter: {
        format(record) {
          return record.level + '<br>'
        },
        formatBatch(records) {
          return records.map(record => this.format(record)).join('\n')
        }
      }
    })
    handler.handleSync(record('debug', 'Test DomHandler'))
    handler.handleSync(record('info', 'Test DomHandler'))

    expect(container.innerHTML).toBe('debug<br>info<br>')
  })

  it('should respect the "level" option', () => {
    const container = document.createElement('div')
    const handler = new DomHandler(container, {
      level: 'notice'
    })

    expect(handler.isHandling('info')).toBe(false)
    expect(handler.isHandling('notice')).toBe(true)

    handler.handleSync(record('info'))
    handler.handleSync(record('notice'))

    expect(container.innerHTML).toMatchSnapshot()
  })

  it('should respect the "bubble" option', () => {
    const bubblingHandler = new DomHandler(document.createElement('div'))
    const nonBubblingHandler = new DomHandler(document.createElement('div'), {
      bubble: false
    })

    expect(bubblingHandler.handleSync(record('debug'))).toBe(false)
    expect(nonBubblingHandler.handleSync(record('debug'))).toBe(true)
  })
})
