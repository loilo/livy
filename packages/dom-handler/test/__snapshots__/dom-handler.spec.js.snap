// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@livy/dom-handler > interaction with "loading" document.readyState > should asynchronously write logs to a selector (with DOM not initially ready) 1`] = `
"
    <div role=\\"listitem\\" style=\\"font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; color: #363844; background: #fafbfd; padding: 0.75em 1em; border-radius: 0.125em; box-shadow: 0 0 0 1px #9195a233, 0 2px #9195a244; margin-bottom: 0.75em\\">
      <time datetime=\\"2015-08-15T14:30:00+02:00\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-bottom: 0.75em\\">2015-08-15 14:30:00</time>

      <span style=\\"position: relative\\">
        <span>Test DomHandler</span>
        <span aria-hidden=\\"true\\" style=\\"position: absolute; left: 0; right: 0; bottom: -1px; height: 8px; background: #9195a211\\"></span>
      </span>

      

      
      
      

      

      <div role=\\"list\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-top: 1.25em\\">
        <span role=\\"listitem\\">
          <span role=\\"label\\" style=\\"display: inline-block; font-weight: inherit\\">channel:</span>
          <span style=\\"display: inline-block\\">logs</span></span>
          <span role=\\"separator\\" style=\\"color: #9195a255\\"> | </span>
          <span role=\\"listitem\\">
          <span style=\\"display: inline-block; font-weight: inherit\\">level:</span>
          <span style=\\"display: inline-block\\">debug</span>
        </span>
      </div>
    </div>"
`;

exports[`@livy/dom-handler > should respect the "level" option 1`] = `
"
    <div role=\\"listitem\\" style=\\"font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; color: #363844; background: #fafbfd; padding: 0.75em 1em; border-radius: 0.125em; box-shadow: 0 0 0 1px #9195a233, 0 2px #08a1ed; margin-bottom: 0.75em\\">
      <time datetime=\\"2015-08-15T14:30:00+02:00\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-bottom: 0.75em\\">2015-08-15 14:30:00</time>

      <span style=\\"position: relative\\">
        <span></span>
        <span aria-hidden=\\"true\\" style=\\"position: absolute; left: 0; right: 0; bottom: -1px; height: 8px; background: #08a1ed11\\"></span>
      </span>

      

      
      
      

      

      <div role=\\"list\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-top: 1.25em\\">
        <span role=\\"listitem\\">
          <span role=\\"label\\" style=\\"display: inline-block; font-weight: inherit\\">channel:</span>
          <span style=\\"display: inline-block\\">logs</span></span>
          <span role=\\"separator\\" style=\\"color: #9195a255\\"> | </span>
          <span role=\\"listitem\\">
          <span style=\\"display: inline-block; font-weight: inherit\\">level:</span>
          <span style=\\"display: inline-block\\">notice</span>
        </span>
      </div>
    </div>"
`;

exports[`@livy/dom-handler > should respect the "reversed" option 1`] = `
"
    <div role=\\"listitem\\" style=\\"font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; color: #363844; background: #fafbfd; padding: 0.75em 1em; border-radius: 0.125em; box-shadow: 0 0 0 1px #9195a233, 0 2px #08a1edaa; margin-bottom: 0.75em\\">
      <time datetime=\\"2015-08-15T14:30:00+02:00\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-bottom: 0.75em\\">2015-08-15 14:30:00</time>

      <span style=\\"position: relative\\">
        <span>Test DomHandler</span>
        <span aria-hidden=\\"true\\" style=\\"position: absolute; left: 0; right: 0; bottom: -1px; height: 8px; background: #08a1ed11\\"></span>
      </span>

      

      
      
      

      

      <div role=\\"list\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-top: 1.25em\\">
        <span role=\\"listitem\\">
          <span role=\\"label\\" style=\\"display: inline-block; font-weight: inherit\\">channel:</span>
          <span style=\\"display: inline-block\\">logs</span></span>
          <span role=\\"separator\\" style=\\"color: #9195a255\\"> | </span>
          <span role=\\"listitem\\">
          <span style=\\"display: inline-block; font-weight: inherit\\">level:</span>
          <span style=\\"display: inline-block\\">info</span>
        </span>
      </div>
    </div>
    <div role=\\"listitem\\" style=\\"font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; color: #363844; background: #fafbfd; padding: 0.75em 1em; border-radius: 0.125em; box-shadow: 0 0 0 1px #9195a233, 0 2px #9195a244; margin-bottom: 0.75em\\">
      <time datetime=\\"2015-08-15T14:30:00+02:00\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-bottom: 0.75em\\">2015-08-15 14:30:00</time>

      <span style=\\"position: relative\\">
        <span>Test DomHandler</span>
        <span aria-hidden=\\"true\\" style=\\"position: absolute; left: 0; right: 0; bottom: -1px; height: 8px; background: #9195a211\\"></span>
      </span>

      

      
      
      

      

      <div role=\\"list\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-top: 1.25em\\">
        <span role=\\"listitem\\">
          <span role=\\"label\\" style=\\"display: inline-block; font-weight: inherit\\">channel:</span>
          <span style=\\"display: inline-block\\">logs</span></span>
          <span role=\\"separator\\" style=\\"color: #9195a255\\"> | </span>
          <span role=\\"listitem\\">
          <span style=\\"display: inline-block; font-weight: inherit\\">level:</span>
          <span style=\\"display: inline-block\\">debug</span>
        </span>
      </div>
    </div>"
`;

exports[`@livy/dom-handler > should write logs to a selector (with DOM ready) 1`] = `
"
    <div role=\\"listitem\\" style=\\"font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; color: #363844; background: #fafbfd; padding: 0.75em 1em; border-radius: 0.125em; box-shadow: 0 0 0 1px #9195a233, 0 2px #9195a244; margin-bottom: 0.75em\\">
      <time datetime=\\"2015-08-15T14:30:00+02:00\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-bottom: 0.75em\\">2015-08-15 14:30:00</time>

      <span style=\\"position: relative\\">
        <span>Test DomHandler</span>
        <span aria-hidden=\\"true\\" style=\\"position: absolute; left: 0; right: 0; bottom: -1px; height: 8px; background: #9195a211\\"></span>
      </span>

      

      
      
      

      

      <div role=\\"list\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-top: 1.25em\\">
        <span role=\\"listitem\\">
          <span role=\\"label\\" style=\\"display: inline-block; font-weight: inherit\\">channel:</span>
          <span style=\\"display: inline-block\\">logs</span></span>
          <span role=\\"separator\\" style=\\"color: #9195a255\\"> | </span>
          <span role=\\"listitem\\">
          <span style=\\"display: inline-block; font-weight: inherit\\">level:</span>
          <span style=\\"display: inline-block\\">debug</span>
        </span>
      </div>
    </div>"
`;

exports[`@livy/dom-handler > should write logs to the referenced element 1`] = `
"
    <div role=\\"listitem\\" style=\\"font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; color: #363844; background: #fafbfd; padding: 0.75em 1em; border-radius: 0.125em; box-shadow: 0 0 0 1px #9195a233, 0 2px #9195a244; margin-bottom: 0.75em\\">
      <time datetime=\\"2015-08-15T14:30:00+02:00\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-bottom: 0.75em\\">2015-08-15 14:30:00</time>

      <span style=\\"position: relative\\">
        <span>Test DomHandler</span>
        <span aria-hidden=\\"true\\" style=\\"position: absolute; left: 0; right: 0; bottom: -1px; height: 8px; background: #9195a211\\"></span>
      </span>

      

      
      
      

      

      <div role=\\"list\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-top: 1.25em\\">
        <span role=\\"listitem\\">
          <span role=\\"label\\" style=\\"display: inline-block; font-weight: inherit\\">channel:</span>
          <span style=\\"display: inline-block\\">logs</span></span>
          <span role=\\"separator\\" style=\\"color: #9195a255\\"> | </span>
          <span role=\\"listitem\\">
          <span style=\\"display: inline-block; font-weight: inherit\\">level:</span>
          <span style=\\"display: inline-block\\">debug</span>
        </span>
      </div>
    </div>
    <div role=\\"listitem\\" style=\\"font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; color: #363844; background: #fafbfd; padding: 0.75em 1em; border-radius: 0.125em; box-shadow: 0 0 0 1px #9195a233, 0 2px #08a1edaa; margin-bottom: 0.75em\\">
      <time datetime=\\"2015-08-15T14:30:00+02:00\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-bottom: 0.75em\\">2015-08-15 14:30:00</time>

      <span style=\\"position: relative\\">
        <span>Test DomHandler</span>
        <span aria-hidden=\\"true\\" style=\\"position: absolute; left: 0; right: 0; bottom: -1px; height: 8px; background: #08a1ed11\\"></span>
      </span>

      

      
      
      

      

      <div role=\\"list\\" style=\\"font-family: SFMono-Regular, Consolas, &quot;Liberation Mono&quot;, Menlo, Courier, monospace; color: #9195a2; display: block; font-size: 0.8em; margin-top: 1.25em\\">
        <span role=\\"listitem\\">
          <span role=\\"label\\" style=\\"display: inline-block; font-weight: inherit\\">channel:</span>
          <span style=\\"display: inline-block\\">logs</span></span>
          <span role=\\"separator\\" style=\\"color: #9195a255\\"> | </span>
          <span role=\\"listitem\\">
          <span style=\\"display: inline-block; font-weight: inherit\\">level:</span>
          <span style=\\"display: inline-block\\">info</span>
        </span>
      </div>
    </div>"
`;
